dev/src/container.rs

```diff
@@ -0,0 +1,197 @@
+            .environment
+            .iter()
+            .map(|(k, v)| format!("{}={}", k, v))
+            .collect();
+
+        let volumes: Vec<String> = self
+            .volumes
+            .iter()
+            .map(|(s, d)| format!("{}:{}", s, d))
+            .collect();
+
+        Config {
+            hostname: Some(self.name.to_string()),
+            env: Some(env),
+            cmd: Some(self.command.clone()),
+            image: Some(self.image.to_string()),
+            host_config: Some(HostConfig {
+                binds: Some(volumes),
+                ..HostConfig::default()
+            }),
+            ..Config::default()
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct ContainerId {
+    name: String,
+    hint: String,
+}
+
+impl ContainerId {
+    pub fn new(name: String, hint: String) -> ContainerId {
+        ContainerId { name, hint }
+    }
+
+    pub fn short_name(&self) -> String {
+        self.name.chars().take(12).collect()
+    }
+}
+
+impl Deref for ContainerId {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        &self.name
+    }
+}
+
+impl Display for ContainerId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} ({})", self.short_name(), self.hint)
+    }
+}
+
+pub struct ContainerResource {
+    docker: Docker,
+    container_id: ContainerId,
+    pub ip: Ipv4Addr,
+    #[allow(unused)]
+    log_handle: JoinHandle<()>,
+}
+
+impl ContainerResource {
+    pub async fn new(spec: &ContainerSpec) -> Result<ContainerResource> {
+        let docker = Docker::connect_with_unix_defaults()?;
+
+        tracing::info!(image_name=%spec.image, "Pulling container.");
+        pull_image(&docker, &spec.image).await?;
+
+        tracing::info!(image_name=%spec.image, "Creating container.");
+        let result = docker
+            .create_container::<String, String>(None, spec.as_config())
+            .await?;
+        let container_id = ContainerId::new(result.id, spec.name.clone());
+
+        tracing::info!(%container_id, "Starting container.");
+        {
+            let options: Option<StartContainerOptions<&str>> = None;
+            docker.start_container(&container_id, options).await?;
+        };
+
+        let result = docker.inspect_container(&container_id, None).await?;
+        let ip: Ipv4Addr = result
+            .network_settings
+            .unwrap()
+            .ip_address
+            .unwrap()
+            .parse()
+            .unwrap();
+
+        tracing::info!(%container_id, %ip, "Container started.");
+
+        let mut log_stream = docker.logs(
+            &container_id,
+            Some(LogsOptions::<&str> {
+                follow: true,
+                stderr: true,
+                stdout: true,
+                timestamps: true,
+                ..LogsOptions::default()
+            }),
+        );
+
+        let mut log_file = File::create(scratch_dir("logs").join(format!("{}.txt", spec.name)))?;
```

Comment:
it'd be nice to be able to pass in a filehandle to ContainerResource (use case here is that we could pass in stdout, which would be nice for just seeing the result)

Alternatively, would be nice to have an option to tee stdout?

Reply:
my use case here is that I quite often write a test then run `ls * | entr -s :TEST_COMMAND`, which reruns that test whenever something changes, and its nice to be able to see the output in that terminal.

Reply:
I don't think the `ContainerResource` is the right level to do that at, maybe it could be done at the test level. I'm struggling to come up with a use case where it's more useful than just running `tail -f` on the file as the test is running, though.

Reply:
Ah, I didn't see your second comment until after mine, I'll give that some thought.